apis/clusters/v1beta1/kafka_webhook_test.go

```diff
@@ -0,0 +1,227 @@
+
+			TestKafkaManifest.Spec.DataCentres = []*KafkaDataCentre{}
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres = kafkaManifest.Spec.DataCentres
+
+			By("changing datacentres fields")
+			prevIntField := kafkaManifest.Spec.DataCentres[0].NodesNumber
+			TestKafkaManifest.Spec.DataCentres[0].NodesNumber++
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].NodesNumber = prevIntField
+
+			prevIntField = kafkaManifest.Spec.DataCentres[0].NodesNumber
+			TestKafkaManifest.Spec.DataCentres[0].NodesNumber = 0
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].NodesNumber = prevIntField
+
+			prevStringField := kafkaManifest.Spec.DataCentres[0].Name
+			TestKafkaManifest.Spec.DataCentres[0].Name += "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].Name = prevStringField
+
+			prevStringField = kafkaManifest.Spec.DataCentres[0].Region
+			TestKafkaManifest.Spec.DataCentres[0].Region += "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].Region = prevStringField
+
+			prevStringField = kafkaManifest.Spec.DataCentres[0].CloudProvider
+			TestKafkaManifest.Spec.DataCentres[0].CloudProvider += "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].CloudProvider = prevStringField
+
+			prevStringField = kafkaManifest.Spec.DataCentres[0].ProviderAccountName
+			TestKafkaManifest.Spec.DataCentres[0].ProviderAccountName += "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].ProviderAccountName = prevStringField
+
+			prevStringField = kafkaManifest.Spec.DataCentres[0].Network
+			TestKafkaManifest.Spec.DataCentres[0].Network += "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].Network = prevStringField
+
+			prevCloudProviderSettings := kafkaManifest.Spec.DataCentres[0].CloudProviderSettings
+			TestKafkaManifest.Spec.DataCentres[0].CloudProviderSettings = []*CloudProviderSettings{prevCloudProviderSettings[0], prevCloudProviderSettings[0]}
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].CloudProviderSettings = prevCloudProviderSettings
+
+			prevMapField := TestKafkaManifest.Spec.DataCentres[0].Tags
+			TestKafkaManifest.Spec.DataCentres[0].Tags["test"] = "test"
+			Expect(k8sClient.Patch(ctx, &TestKafkaManifest, patch)).ShouldNot(Succeed())
+			TestKafkaManifest.Spec.DataCentres[0].Tags = prevMapField
```

Comment:
Map is a reference type so they should be identical when you returns the previous map state. You should create a new map and assign it to the test kafka manifest.

Reply:
agree, and added other cases